packages/gatsby/src/schema/node-model.js

```diff
@@ -521,10 +521,10 @@ class LocalNodeModel {
     let iterations = 0
     let ids = this._rootNodeMap.get(obj)
     if (!ids) {
-      ids = []
+      ids = new Set()
     }
     if (obj?.parent) {
```

Comment:
While this is not something that is broken by changes - I was looking at steps to reproduce mentioned in https://github.com/sdobz/gatsby-markdown-frontmatter-repro
> Create a markdown page with "parent" in the frontmatter

the `parent` field here is not meant to be something that represent "data" and instead is meant for parent node id ( https://www.gatsbyjs.com/docs/reference/graphql-data-layer/node-interface/#parent )

While we can't exactly check if `obj` here is a "node", we can add some type checking at least to minimize potentially using unexpected data type and later if this `parent` string value doesn't represent actual parent id - `getNodeById` call on line 533 will just return `null` most likely

```suggestion
    if (obj?.parent && typeof obj.parent === 'string') {
```

The changes themselves (array => set) make sense and seems needed 👍

Reply:
Hmn, my lack of understanding of the gatsby data model is showing. Why is `findRootNodeAncestor` being run on "frontmatter" - an arbitrary object on a Node?

```javascript
  findRootNodeAncestor(obj, predicate = null) {
    console.log("findRootNodeAncestor", obj)
    let iterations = 0
```

```
findRootNodeAncestor { title: '', parent: 'defined', thumbnail: './thumbnail.png' }
```

It's a field resolver `fileByPath`... What's a resolver, and why is it sniffing at this object? Ah! GraphQL term. It needs to do _things_ to that thumbnail field in order to file-ify it.

Trying to figure out how to... NOT treat maybe-a-node things as a node I'd probably invert this and maintain a registry of object->parent node that is recursively populated on node construction. There's likely a good reason this isn't done (are nodes mutated? constructed on demand in a way that they don't have consistent references?) so yea leaving this as good enough for now.